Fix support for Internal STM32F4x flash and add STM32L4x support

superna9999 opened the pull request:
This PR adds support for the dual-bank STM32L4x Flash Controller along fixes to the STM32F4x flash driver.

The STM32L4x SoCs embeds a slightly different embedded flash controller
from the STM32F4x SoCs.

This particular controller has the following properties :
 - Up to 2 512KiB banks divided in 2KiB pages
 - Flash can be accessed in any sizes
 - Flash must be written in 64bit aligned 64bit double-words

The drivers/flash/flash_stm32f4x.c is refactored into a new common
drivers/flash/flash_stm32.c and drivers/flash/flash_stm32l4x.c is created
with the STM32L4x specific functions.

To ease the refactoring and keep common functions, the STM32L4x flash
headers are slightly modified to match the hardware reference naming and
solve compilation issues.

superna9999 committed:
flash: Rework and add flash device support for STM32L4x SoCs

The STM32L4x SoCs embeds a slightly different embedded flash controller
from the STM32F4x SoCs.

This particular controller has the following properties :
 - Up to 2 512KiB banks divided in 2KiB pages
 - Flash can be accessed in any sizes
 - Flash must be written in 64bit aligned 64bit double-words

The drivers/flash/flash_stm32f4x.c is refactored into a new common
drivers/flash/flash_stm32.c and drivers/flash/flash_stm32l4x.c is
created with the STM32L4x specific functions.

To ease the refactoring and keep common functions, the STM32L4x flash
headers are slightly modified to match the hardware reference naming
and solve compilation issues.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

galak reviewed (changes requested):
Can we split the flash test out of this PR.  It feels like we should have a more generic flash test that could be used on any device, not sure what all is specific to the L4x controller.

galak reviewed (commented):
How much does this really differ from flash_stm32f4x.c?  Can we refactor that code to be used by both f4 & l4?

superna9999 commented:
Hi,

OK to split out the flash test, but on L4x, write data must be 8bytes length and aligned, so the test must be generic enough to take this in account.

For the common driver, L4x differ from STM32F4x on:
- L4x cannot write bytes, only double words, ECC is on each double-word so next write on this double-word will fail
- L4x pages are 2k on a dual bank, really different from the F4x

But we could factor out all the boilerplate code to only have really specific F4x/L4x code in separate files.

galak commented:
Lets at least factor out the common code into .h or .c

superna9999 commented:
I merged the two drivers in a single common "flash_stm32" driver, but I check the F4x builds but I don't have any F4x supported board to test this out.

superna9999 commented:
In case you wonder, I made an experimental (non tested) merge of the F3x driver into this common scheme:
https://github.com/superna9999/zephyr/commit/a662ea3374dc24d6b22c9307ca6bdd6e492a1b9a

erwango reviewed (changes requested):
First comments

superna9999 commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,161 @@
+
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
+#define FLASH_KEY1	((uint32_t)0x45670123U) /* Flash key1 */
+#define FLASH_KEY2	((uint32_t)0xCDEF89ABU)	/* Flash key2 */
```
Yes, but not for L4x:

> $ grep FLASH_KEY ext/hal/st/stm32cube/stm32f*/soc/ -R | wc -l
> 336

They are defined in the stm32f*****.h files.

> $ grep FLASH_KEY ext/hal/st/stm32cube/stm32l*/soc/ -R | wc -l
> 0

They are defined in stm32cube/stm32l4xx/drivers/include/stm32l4xx_hal_flash.h

How should that be handled ?

erwango replied:
F4 and F3 use them already, so you should not need much.
I think if you include soc.h, you should get it.

superna9999 replied:
Sorry I don't understand since it's not declared in the soc files for L4x

superna9999 replied:
Indeed, but I don't see any warning when I redefine them.

erwango replied:
I'll have a look.
Btw, thanks for this factorizing work. I can test F4 and F3 if needed.

superna9999 replied:
Thanks, I rebased the F3x merge code on this updated PR: https://github.com/superna9999/zephyr/tree/flash_stm32f3x_merge
if you want to test on a F3.

superna9999 commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,161 @@
+#define FLASH_KEY1	((uint32_t)0x45670123U) /* Flash key1 */
+#define FLASH_KEY2	((uint32_t)0xCDEF89ABU)	/* Flash key2 */
+
+#define STM32L4X_FLASH_TIMEOUT	((u32_t) 0x000B0000)
```
Yes, it's also the same for F3x

superna9999 commented on arch/arm/soc/st_stm32/stm32f4/flash_registers.h:
```diff
@@ -32,10 +32,10 @@ union __flash_acr {
 	volatile union __flash_acr acr;
-	volatile u32_t key;
+	volatile u32_t keyr;
 	volatile u32_t optkey;
```
Ok

superna9999 commented on arch/arm/soc/st_stm32/stm32f4/flash_registers.h:
```diff
@@ -32,10 +32,10 @@ union __flash_acr {
-	volatile u32_t ctrl;
+	volatile u32_t sr;
+	volatile u32_t cr;
 	volatile u32_t optctrl;
```
Ok

superna9999 commented:
While testing on a STM32F407 SoC, I spotted some issues with the original code for STM32F4x:
- MPU for flash is forced to RO, a config to at least switch flash to P_RW_U_NA should be added
- All registers were accessed without volatile (fixed in this merge)
- All bits clear in write_byte() and erase_sector() were wrong:
`regs->cr &= (FLASH_CR_SER | FLASH_CR_SNB);`
instead of:
`regs->cr &= ~(FLASH_CR_SER | FLASH_CR_SNB);`

Should I push all the F4x fixes in a separate commit, before the merge of after ?

superna9999 committed:
flash_stm32f4: RDERR is not present on STM32F407

The RDERR bit is not present on F4x5 and F4x7 SoCs, so only use it when
available in the HAL.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
flash_stm32f4: bits were badly unset in erase and write functions

A typo was present ion the erase and write function were bit were
uncorrectly unset but all other bit except the mask were unset.
The code still worked because the two typos fixed themselves.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

erwango commented:
Adding @ldts and @fvincenzo to comment on original F4 driver and MPU code

ldts commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -79,7 +79,7 @@ static int write_byte(off_t offset, u8_t val, struct stm32f4x_flash *regs)
 	}
 
-	regs->ctrl &= CR_PSIZE_MASK;
+	regs->ctrl &= ~CR_PSIZE_MASK;
```
ACK

ldts commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -117,7 +117,7 @@ static int erase_sector(u16_t sector, struct stm32f4x_flash *regs)
 
 	rc = wait_flash_idle(regs);
-	regs->ctrl &= (FLASH_CR_SER | FLASH_CR_SNB);
+	regs->ctrl &= ~(FLASH_CR_SER | FLASH_CR_SNB);
```
ACK

superna9999 commented:
@fvincenzo My MPU patch is kind of a duplicate of your https://github.com/fvincenzo/zephyr/commit/9f757cc91b6a5458f819f1e657c70e1f79e5244b

galak reviewed (changes requested):
Lets pull the MPU bits into a separate PR

ldts commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,155 @@
+int flash_stm32_wait_flash_idle(struct flash_stm32_priv *p)
+{
+#if defined(CONFIG_SOC_SERIES_STM32F4X)
+	volatile struct stm32f4x_flash *regs = p->regs;
```
the register fields in the structure already carry the volatile keyword so I believe it is redundant  doing it again. but if you need to do this, out of curiosity, what compiler are you using?...seems to me a compiler issue...

superna9999 replied:
The volatile is not present in the STM32L4x definitions.

This should be synchronized among the STM32 headers, where should "volatile" be added ?

erwango replied:
+1 (removing my redundant comment).
And if this is the right explanation, it would be worse aligning the structures one way or the other.

ldts replied:
I suppose for consistency you should either add them to the stm32l4x definitions or remove them from the stm32f4x; since stm32f4x is already upstream I believe the better thing to do is to add them to stm32l4x.

superna9999 replied:
Done, thanks.

superna9999 commented:
@galak Right, I didn't realize @fvincenzo was working on something similar

superna9999 commented on arch/arm/soc/st_stm32/stm32f4/flash_map.h:
```diff
@@ -41,7 +41,8 @@ struct stm32f4x_flash_sector stm32f4xx_sectors[] = {
 	(stm32f4xx_sectors[ARRAY_SIZE(stm32f4xx_sectors) - 1].end)
 
-int stm32f4x_get_sector(off_t offset)
+__attribute__((unused))
```
Done

erwango commented:
About "All registers were accessed without volatile (fixed in this merge)", isn't it because flash_registers were defined as "volatile u32_t" on f4 (while defined as u32_t on L4) ?
If this is linked, maybe you could align structure definitions.

galak commented:
@erwango you good with this version?

erwango commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -0,0 +1,196 @@
+	return rc;
+}
+
+int flash_stm32_erase(struct device *dev, off_t offset, size_t len)
```
Only difference in this function is the ...get_page vs ...get_sector used in F4.
While I agree there is a real difference behind these words, providing a generic API would help factorizing a bit further and reduce differences between flash_stm32l4x.c vs flash_stm32f4x.c down to the minimum HW specificcs (length of erase/write operation, ECC availability, ....
Can you have a try?

superna9999 replied:
Ok, can you have a look of what I did ?

erwango commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -0,0 +1,196 @@
+	return rc;
+}
+
+int flash_stm32_write(struct device *dev, off_t offset,
```
Same comment as on flash_stm32_erase

superna9999 replied:
Same here

erwango reviewed (changes requested):
Thanks Neil for the factorization effort.
Some comments to push it further to the essential

superna9999 commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -0,0 +1,196 @@
+	return offset >> STM32L4X_PAGE_SHIFT;
+}
+
+int flash_stm32_check_status(struct flash_stm32_priv *p)
```
Right, I need to check for F3, but it's OK.

The only tweak is to add a:
#if defined(FLASH_FLAG_PGPERR)
along the:
#if defined(FLASH_FLAG_RDERR)

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,265 @@
+			break;
+		}
+	}
+#else
```
I should admit the result is not as simple as I'd have expected, specially if we need to define additional CONFIG_ switches.
What about introducing a new generic function flash_stm32_block_erase_loop(offset, p, len)?
Then, you can have get/erase_page and get/erase_sector in respective f4 and l4 files.

superna9999 replied:
Ok, I see now, let me start again

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,265 @@
+#elif defined(CONFIG_STM32_FLASH_WRITE_DOUBLE_WORD)
+		rc = flash_stm32_write_dword(offset,
+					     ((const u64_t *) data)[i], p);
+#endif
```
same as before with  flash_stm32_block_erase_loop(offset, data, p)

superna9999 replied:
Hmm, OK I understand, I'll try again ;-)

erwango reviewed (changes requested):
Thanks for the quick feedback. Though, this is not as generic as I was expecting, and I'm not quite fond of introducing new config flags, while series information should suffice.
New proposal to enhance main file genericity while keeping the "page/sector" "word/byte" dinstinction in f4/l4 specific files.

erwango commented on drivers/flash/Kconfig.stm32fxx:
```diff
@@ -10,10 +10,10 @@ if FLASH && SOC_FAMILY_STM32
 	bool
 	prompt "STM32 flash driver"
-	depends on (SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X)
+	depends on (SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X)
```
Update file name and header to fit new scope

superna9999 replied:
Done

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,229 @@
+}
+
+static int flash_stm32_read(struct device *dev, off_t offset, void *data,
+	size_t len)
```
fix alignment here

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -0,0 +1,229 @@
+	POST_KERNEL,
+	CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+	&flash_stm32_api);
+
```
Fix alignment on the whole macro

erwango reviewed (changes requested):
Quite in line with what I had in mind, thanks!
Few minors comments, and it will LGTM

galak committed:
Merge branch 'arm' into flash_stm32lx

erwango reviewed (approved):
Good work, thanks

galak committed:
Merge branch 'arm' into flash_stm32lx
